<html>
<head>
    <title>??? Manual</title>
    <style type="text/css">
        .code {
            background-color: lightgray;
        }
    </style>
</head>
<body>
<h1>??? Manual</h1>

1. <a href="#sec1">About ???</a><br>
2. <a href="#sec2">Installation</a><br>
3. <a href="#sec3">Running ???</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.1. <a href="#sec3.1">RNA scaffolder</a><br>

&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.1.1. <a href="#sec3.1.1">Input</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.1.2. <a href="#sec3.1.2">Command line options</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.1.3. <a href="#sec3.1.3">Output</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.1.4. <a href="#sec3.1.4">Example</a><br>

&nbsp;&nbsp;&nbsp;&nbsp;3.2. <a href="#sec3.2">Graph visualization</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.2.1. <a href="#sec3.2.1">Input</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.2.2. <a href="#sec3.2.2">Command line options</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.2.3. <a href="#sec3.2.3">Output</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.2.4. <a href="#sec3.2.4">Example</a><br>
4. <a href="#sec4">References</a><br>
5. <a href="#sec5">Feedback and bug reports</a><br>
<br>

<a name="sec1"></a>
<h2>1. About ???</h2>

<a name="sec2"></a>
<h2>2. Installation</h2>
Downloading and compiling source code.

<p>
    If you wish to compile ??? by yourself you will need the following libraries to be pre-installed:
<ul>
    <li>g++ </li>
    <li>cmake </li>
    <li>zlib</li>
    <li>Threads</li>
    <li>Boost</li>
    <li>bowtie2</li>
    <li>STAR</li>
    <li>nucmer</li>
</ul>

<p>
    If you meet these requirements, you can download the ???:

<pre class="code">
<code>
    wget ???
    tar -xzf ???
    cd ???
</code>
</pre>

<p>
    and build it with the following script:

<pre class="code">
<code>
    ./compile.sh
</code>
</pre>

<p>
    ??? will be built in the directory <code>./bin</code>. If you wish to install ??? into another directory, you can specify full path of destination folder by running the following command in <code>bash</code> or <code>sh</code>:

<pre class="code">
<code>
    PREFIX=&lt;destination_dir> ./compile.sh
</code>
</pre>

<p>
    for example:

<pre class="code">
<code>
    PREFIX=/usr/local ./compile.sh
</code>
</pre>

<p>
    which will install ??? into <code>/usr/local/bin</code>.

<p>
    After installation you will get the same files in <code>./bin</code> (or <code>&lt;destination_dir>/bin</code> if you specified PREFIX) directory:
<ul>
    <li><code>rna_scaffolder.py</code>  (main executable script for build rna scaffolders)</li>
    <li><code>visualize_scaffold_graph.py</code>  (main executable script for visualize scaffold graph) </li>
    <li><code>build</code>  (build graph module)</li>
    <li><code>filter</code>  (filter and print graph module)</li>
    <li><code>addInfoToGraph</code>  (add scaffolds to graph module) </li>
    <li><code>addBothPath</code>  (add scaffolds to graph module) </li>
    <li><code>mergeGraph</code>  (merge graph module)</li>
</ul>

<p>
    We also suggest adding ??? installation directory to the <code>PATH</code> variable.

<a name="sec3"></a>
<h2>3. Running ???</h2>

<a name="sec3.1"></a>
<h3>3.1. RNA scaffolder</h3>

<a name="sec3.1.1"></a>
<h4>3.1.1. Input </h4>

<p>
    ??? takes as input contigs in FASTA format and paired-end RNA reads or single (unpaired) RNA reads in FASTA and FASTQ.
<p>

<a name="sec3.1.2"></a>
<h4>3.1.2. Command line options </h4>

<p>
    To run RNA scaffolder from the command line, type

<pre class="code">
<code>
    rna_scaffolder.py [options]
</code>
</pre>
Note that we assume that ??? installation directory is added to the <code>PATH</code> variable (provide full path to ??? executable otherwise: <code>&lt;installation dir>/rna_scaffolder.py</code>).

<a name="basicopt"></a>
<h4>Basic options</h4>

<p>
    <code>-c</code> (or <code>--contig</code>) <code> &lt;file_name> </code>><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with contigs in fasta format. Required option.
</p>

<p>
    <code>--rna-p &lt;file_name_1>  &lt;file_name_2> </code>
    &nbsp;&nbsp;&nbsp;&nbsp; File with left reads and file with right eads for paired-end library.
</p>

<p>
    <code>--rna-s &lt;file_name> </code>
    &nbsp;&nbsp;&nbsp;&nbsp; File for single-read library.
</p>

<p>
    <code>-o</code> (or <code>--local_output_dir</code>) <code> &lt;output_dir> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Specify the output directory. The default output dir is  "./"
</p>


<a name="sec3.1.3"></a>
<h4>3.1.3. Output</h4>

??? stores all output files in <code>&lt;output_dir> </code>, which is set by the user.

<ul>
    <li><code>&lt;output_dir>/scaffolds.fa</code> contains resulting scaffolds </li>
    <li><code>&lt;output_dir>/graph.gr</code> contig graph </li>
    <li><code>&lt;output_dir>/out.info</code> scaffolds description</li>
</ul>

<p>
    Scaffolds names in ??? output FASTA files have the following format: <br><code>path179 </code><br> Here <code>179</code> is the number of the scaffold.
</p>


<a name="sec3.1.4"></a>
<h4>3.1.4. Example</h4>


<a name="sec3.2"></a>
<h3>3.2. Graph visualization</h3>
<a name="sec3.2.1"></a>
<h4>3.2.1. Input </h4>

<a name="sec3.2.2"></a>
<h4>3.2.2. Command line options </h4>
<p>
    To run scaffold graph visualization from the command line, type

<pre class="code">
<code>
    visualize_scaffold_graph.py [options]
</code>
</pre>
Note that we assume that ??? installation directory is added to the <code>PATH</code> variable (provide full path to ??? executable otherwise: <code>&lt;installation dir>/visualize_scaffold_graph.py</code>).

<a name="basicopt"></a>
<h4>Basic options</h4>
<p>
    <code>-c</code> (or <code>--contig</code>) <code> &lt;file_name> </code>><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with contigs in fasta format. Required option.
</p>
<p>
    <code>--rna-p &lt;file_name_1 file_name_2> </code>
    &nbsp;&nbsp;&nbsp;&nbsp; File with left reads and file with right reads for paired-end RNA library.
</p>
<p>
    <code>--rna-s &lt;file_name> </code>
    &nbsp;&nbsp;&nbsp;&nbsp; File for single-read RNA library.
</p>
<p>
    <code>--dna-p &lt;file_name_1 file_name_2> </code>
    &nbsp;&nbsp;&nbsp;&nbsp; File with left reads and file with right reads for paired-end DNA library.
</p>
<p>
    <code>--ref &lt;file_name> </code>
    &nbsp;&nbsp;&nbsp;&nbsp; File with reference.
</p>
<p>
    <code>--scafinfo &lt;file_name> </code>
    &nbsp;&nbsp;&nbsp;&nbsp; File with scaffolds in info format.
</p>
<p>
    <code>--scafpath &lt;file_name> </code>
    &nbsp;&nbsp;&nbsp;&nbsp; File with scaffolds in path format.
</p>
<p>
    <code>--label &lt;label1 label2 ...></code>
    &nbsp;&nbsp;&nbsp;&nbsp; Labels for libraries in given order.
</p>
<p>
    <code>--color &lt;color1 color2 ...></code>
    &nbsp;&nbsp;&nbsp;&nbsp; Colors for libraries in given order.
</p>
<p>
    <code>-o</code> (or <code>--local_output_dir</code>) <code> &lt;output_dir> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Specify the output directory. The default output dir is  "./"
</p>

<a name="sec3.2.3"></a>
<h4>3.2.3. Output</h4>

??? stores all output files in <code>&lt;output_dir> </code>, which is set by the user.

<ul>
    <li><code>&lt;output_dir>/outdot/</code>  dot files with graph</li>
    <li><code>&lt;output_dir>/graph.gr</code>  graph description </li>
    <li><code>&lt;output_dir>/graph.ps</code> picture with graph </li>
</ul>


<a name="sec3.2.4"></a>
<h4>3.2.4. Example</h4>


<a name="sec4"></a>
<h2>4. References </h2>

<a name="sec5"></a>
<h2>5. Feedback and bug reports </h2>

<br/>
</body>
</html>