<html>
<head>
    <title>SGTK Manual</title>
    <style type="text/css">
        .code {
            background-color: lightgray;
        }
    </style>
</head>
<h1>SGTK Manual</h1>

1. <a href="#sec1">About SGTK</a><br>
2. <a href="#sec2">Installation</a><br>
3. <a href="#sec3">Running SGTK</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.1. <a href="#sec3.1">Graph visualization</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.1.1. <a href="#sec3.1.1">Input</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.1.2. <a href="#sec3.1.2">Command line options</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.1.3. <a href="#sec3.1.3">Output</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.1.4. <a href="#sec3.1.4">Example</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;3.2. <a href="#sec3.1">RNA scaffolder</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.2.1. <a href="#sec3.2.1">Input</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.2.2. <a href="#sec3.2.2">Command line options</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.2.3. <a href="#sec3.2.3">Output</a><br>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;3.2.4. <a href="#sec3.2.4">Example</a><br>
4. <a href="#sec4">References</a><br>
5. <a href="#sec5">Feedback and bug reports</a><br>
<br>

<a name="sec1"></a>
<h2>1. About SGTK</h2>
<p>
    SGTK &ndash; Scaffold Graph ToolKit &ndash; is a tool which is capable of scaffold graph construction,
    interactive visualization and building scaffolds by RNA-seq data. This manual will help you to install and run SGTK.
<p>
    The current version of SGTK takes as input a set of contigs and an arbitrary number of linkage information sources, such as:
<ul>
    <li>paired reads</li>
    <li>long reads</li>
    <li>paired and unpaired RNA-seq reads</li>
    <li>scaffolds</li>
    <li>assembly graph in GFA, FASTG formats</li>
    <li>reference sequences</li>
</ul>

SGTK works with Illumina RNA-seq paired-end and unpaired reads for building scaffolds.

<a name="sec2"></a>
<h2>2. Installation</h2>
To install SGTK download the source code and compile it yourself.

<p>
    To compile SGTK by yourself you will need the following libraries to be pre-installed:
<ul>
    <li>python3</li>
    <li>g++ </li>
    <li>cmake </li>
    <li>zlib</li>
    <li>Threads</li>
    <li>Boost</li>
    <li>minimap2</li>
    <li>STAR</li>
    <li>nucmer</li>
</ul>

<p>
    If you meet these requirements, you can download the SGTK source code:

<pre class="code">
<code>
   git clone https://github.com/olga24912/SGTK.git
   cd SGTK/scaffolder/
</code>
</pre>

<p>
    and build it with the following script:

<pre class="code">
<code>
   ./compile.sh
</code>
</pre>

<p>
    SGTK will be built in the directory <code>./bin</code>. If you wish to install SGTK into another directory, you can specify the full path of destination folder by running the following command in <code>bash</code> or <code>sh</code>:

<pre class="code">
<code>
   PREFIX=&lt;destination_dir> ./compile.sh
</code>
</pre>

<p>
    for example:

<pre class="code">
<code>
   PREFIX=/usr/local ./compile.sh
</code>
</pre>

<p>
    which will install SGTK into <code>/usr/local/bin</code>.

<p>
    After installation you will get the following files in <code>./bin</code> (or <code>&lt;destination_dir>/bin</code> if you specified PREFIX) directory:
<ul>
    <li><code>rna_scaffolder.py</code>  (main executable script for building rna scaffolders)</li>
    <li><code>visualize_scaffold_graph.py</code>  (main executable script for visualization scaffold graph) </li>
    <li><code>build</code>  (graph construction module)</li>
    <li><code>filter</code>  (graph simplification and building scaffolds module)</li>
    <li><code>mergeGraph</code>  (graph merging module)</li>
    <li><code>readSplitter</code> (module for splitting RNA-seq reads)</li>

    <li><code>mainPage.html</code> (main html page for visualization)</li>
    <li><code>scripts/</code> (folder containing JS necessary for visualization)</li>
    <li><code>scripts/script.js</code> </li>
    <li><code>scripts/light.css</code> </li>
    <li><code>scripts/search.js</code> </li>
    <li><code>scripts/showListScript.js</code> </li>
    <li><code>scripts/zoomChoose.js</code> </li>
    <li><code>scripts/scaffoldgraph.js</code> </li>
    <li><code>scripts/Icon.png</code> </li>
    <li><code>scripts/defaultWeight.js</code> </li>
    <li><code>scripts/findAreaScript.js</code> </li>
    <li><code>scripts/freeLayout.js</code> </li>
    <li><code>scripts/handleScaffolds.js</code> </li>
    <li><code>scripts/handleAlongChromosomes.js</code> </li>
    <li><code>scripts/handleDiffInLibs.js</code> </li>
    <li><code>scripts/external/cytoscape.js</code></li>
    <li><code>scripts/external/cytoscape-dagre.js</code></li>
    <li><code>scripts/external/dagre.js</code></li>
    <li><code>scripts/external/cytoscape-qtip.js</code></li>
    <li><code>scripts/external/cytoscape.js-navigator/bower.json</code></li>
    <li><code>scripts/external/cytoscape.js-navigator/package.json</code></li>
    <li><code>scripts/external/cytoscape.js-navigator/cytoscape-navigator.js</code></li>
    <li><code>scripts/external/cytoscape.js-navigator/gulpfile.js</code></li>
</ul>

<p>
    We also suggest adding SGTK installation directory to the <code>PATH</code> variable.

    <a name="sec3"></a>
<h2>3. Running SGTK</h2>
<a name="sec3.1"></a>
<h3>3.1. Graph visualization</h3>
<a name="sec3.1.1"></a>
<h4>3.1.1. Input </h4>
SGTK takes as input contigs, reference and an arbitrary number of linkage
information sources.
<h4> Contigs </h4>
Contigs in FASTA format. You can provide a few contigs files, in this case they will be merged together.
Make sure that all contigs have different names.

If you have contigs in forward and reverse-complementary copies, provide only one copy.
<h4> Reference </h4>
If you provide reference genome in FASTA format the genome browser visualization will be available.
In case of providing several files they will be merged together and chromosomes’ names will be changed
depending on the references’ files names.

If you have chromosomes in forward and reverse-complementary copies, provide only one copy.

<h4> Paired-end and mate-pair reads</h4>
SGTK takes paired reads in FASTA or FASTQ format in two files or alignments in SAM/BAM format in two files as well.
In case of providing reads in FASTA/FASTQ format they will be aligned on contigs by minimap2:
<pre class="code">
   <code>
       minimap2 -ax sr &lt;contigs_file> &lt;dna1> > dna1.sam
       minimap2 -ax sr &lt;contigs_file> &lt;dna2> > dna2.sam
   </code></pre>

<h4> RNA-seq reads </h4>
You can provide both paired and single RNA-seq reads. Paired reads will be aligned to the contigs by STAR independently. Single reads will be split into two parts and after that will be aligned by STAR.

<h4> PacBio reads </h4>
SGTK takes PacBio reads in FASTA or FASTQ format and aligns them on contigs by minimap2:
<pre class="code">
   <code>
       minimap2 -x map-pb &lt;contigs_file> &lt;pacbio> > out.paf
   </code></pre>

<h4> Scaffolds </h4>
There are two ways to provide scaffolds:
<ul>
    <li>scaffolds in FASTA format. In this case contigs will be aligned to scaffolds by nucmer.</li>
    <li>scaffolds paths in INFO format </li>
</ul>
In INFO format each line describes scaffold in the following format:<br>
<pre class="code">
   <code>
       >SCAFFOLD_NAME (CONTIG_NAME_0 CONTIG_ID_0 DIRECTION_0) (CONTIG_NAME_1 CONTIG_ID_1 DIRECTION_1)
   </code></pre>

For example, one line in INFO format:
<pre class="code">
   <code>
       >scaffoldName (contigName0 0 +) (contigName1 1 -) (contigName2 2 -) (contigName3 3 +)
   </code></pre>

<h4> Connection list </h4>
Also it is possible to provide a file with your own connections, where each line describes one connection in the following format:
<pre class="code">
       <code>
           (CONTIG_NAME_0 CONTOG_DIRECTION_0) (CONTIG_NAME_1 CONTOG_DIRECTION_1) WEIGHT LEN "EXTRA_INFO"
       </code></pre>
For example:
<pre class="code">
       <code>
           (contig0 -) (contig1 +) 32.5 1168 "some info about connection"
       </code></pre>


<h4> Scaffold Graph </h4>
All information about connections will be merged in one scaffold graph in internal format. You can provide
extra scaffold graph on your own.

Scaffold graph has the following format:
<pre class="code">
       <code>
           LIB_DESCRIPTION
           NODES_DESCRIPTION
           EDGES_DESCRIPTION
       </code></pre>

The format for LIB_DESCRIPTION:
<pre class="code">
       <code>
           LIB_NUMBER
           (l LIB_ID LIB_COLOR LIB_NAME LIB_TYPE)*LIB_NUMBER
       </code></pre>
Where LIB_TYPE = {CONNECTION | PACBIO | DNA_PAIR | RNA_PAIR | RNA_SPLIT_50 | RNA_SPLIT_30}
<br>
<br>
The format for NODES_DESCRIPTION:
<pre class="code">
       <code>
           NODES_NUMBER
           (v NODE_ID NODE_NAME NODE_LEN)* NODES_NUMBER
       </code></pre>

Format for EDGES_DESCRIPTION:
<pre class="code">
       <code>
           EDGES_NUMBER
           (e EDGE_ID NODE_ID_0 NODE_ID_1 LIB_ID WEIGHT LEN "EXTRA_INFO")*EDGES_NUMBER
       </code></pre>

<br>
For example:
<pre class = "code">
       <code>
           1
           l 0 #ff0000 lib_name DNA_PAIR
           4
           v 0 node_0 20132
           v 1 node_0-rev 20132
           v 2 node_1 20400
           v 3 node_1-rev 20400
           2
           e 0 0 2 0 32.5 200 "some extra info"
           e 1 3 1 0 32.5 200
       </code></pre>

<a name="sec3.1.2"></a>
<h4>3.1.2. Command line options </h4>
<p>
    To run scaffold graph visualization from the command line, type

<pre class="code">
<code>
   visualize_scaffold_graph.py [options]
</code>
</pre>
Note that we assume that SGTK installation directory is added to the <code>PATH</code> variable (otherwise provide full path to SGTK executable: <code>&lt;installation dir>/visualize_scaffold_graph.py</code>).

<a name="basicopt"></a>
<h4>Options</h4>
<p>
    <code>-h</code> (or <code>--help</code>)
    &nbsp;&nbsp;&nbsp;&nbsp;Print help.
</p>
<p>
    <code>-c</code> (or <code>--contig</code>) <code> &lt;file_name> </code>
    &nbsp;&nbsp;&nbsp;&nbsp;File with contigs in fasta format. Required option.
</p>
<p>
    <code>-s</code> (or <code>--scaffolds</code>) <code> &lt;file_name> </code>
    &nbsp;&nbsp;&nbsp;&nbsp;File with scaffolds in fasta format.
</p>
<p>
    <code>--fr &lt;file_name_1> &lt;file_name_2> </code>
    &nbsp;&nbsp;&nbsp;&nbsp; File with left reads and file with right reads for paired-end/mate-pair DNA library with forward-reverse orientation.
</p>
<p>
    <code>--rf &lt;file_name_1> &lt;file_name_2> </code>
    &nbsp;&nbsp;&nbsp;&nbsp; File with left reads and file with right reads for paired-end/mate-pair DNA library with reverse-forward orientation.
</p>
<p>
    <code>--ff &lt;file_name_1> &lt;file_name_2> </code>
    &nbsp;&nbsp;&nbsp;&nbsp; File with left reads and file with right reads for paired-end/mate-pair DNA library with forward-forward orientation.
</p>
<p>
    <code>--pacbio &lt;file_name> </code>
    &nbsp;&nbsp;&nbsp;&nbsp; File with PacBio reads.
</p>
<p>
    <code>--scg &lt;file_name> </code>
    &nbsp;&nbsp;&nbsp;&nbsp;File with connection list.
</p>
<p>
    <code>--gr &lt;file_name> </code>
    &nbsp;&nbsp;&nbsp;&nbsp;File with scaffold graph description.
</p>
<p>
    <code>--rna-p &lt;file_name_1> &lt;file_name_2> </code>
    &nbsp;&nbsp;&nbsp;&nbsp; File with left reads and file with right reads for paired-end RNA library.
</p>
<p>
    <code>--rna-s &lt;file_name> </code>
    &nbsp;&nbsp;&nbsp;&nbsp; File for single-read RNA library.
</p>
<p>
    <code>--ref &lt;file_name> </code>
    &nbsp;&nbsp;&nbsp;&nbsp; File with reference in FASTA format.
</p>
<p>
    <code>--scafinfo &lt;file_name> </code>
    &nbsp;&nbsp;&nbsp;&nbsp; File with scaffolds in info format.
</p>
<p>
    <code>--label &lt;label1 label2 ...></code>
    &nbsp;&nbsp;&nbsp;&nbsp; Labels for libraries in given order.
</p>
<p>
    <code>--color &lt;color1 color2 ...></code>
    &nbsp;&nbsp;&nbsp;&nbsp; Colors for libraries in given order.
</p>
<p>
    <code>-o</code> (or <code>--local_output_dir</code>) <code> &lt;output_dir> </code>
    &nbsp;&nbsp;&nbsp;&nbsp;Specify the output directory. The default output directory is  "./"
</p>

<a name="sec3.1.3"></a>
<h4>3.1.3. Output</h4>

SGTK stores all output files in <code>&lt;output_dir> </code>, which is set by the user.

<ul>
    <li><code>&lt;output_dir>/scripts/</code> files with scaffold graph description. It is needed for main.html to work properly </li>
    <li><code>&lt;output_dir>/main.html</code> main file for graph visualization </li>
</ul>

Run main.html to see visualisation.

<a name="sec3.1.4"></a>
<h4>3.1.4. Example</h4>
<p>
    To test the toy data set, you can also run the following command from the SGTK <code>bin</code> directory:

<pre class="code">
<code>
   python3 visualize_scaffold_graph.py -c ../share/test_dataset/contigs.fasta --rna-p ../share/test_dataset/read_1.fasta \
   ../share/test_dataset/read_2.fasta -o test
</code>
</pre>

<p>
    If you have several paired-end reads, scaffolds and reference:
<p>
    <li> contigs
        <pre class="code">
   <code>
   ../share/test_dataset/contigs.fasta
   </code>
   </pre>

    <li> paired-end library
        <pre class="code">
<code>
   ../share/test_dataset/read_1.fasta
   ../share/test_dataset/read_2.fasta
</code>
</pre>
<p>
    <li> scaffolds

        <pre class="code">
<code>
   scaf.info
</code>
</pre>

<p>
    <li> reference
        <pre class="code">
<code>
   ref.fasta
</code>
</pre>
<p>

    If you would like to set labels and colors, you need to set labels and colors for all libraries in order of definition
<pre class="code">
<code>
   python3 visualize_scaffold_graph.py -c ../share/test_dataset/contigs.fasta \
   --fr ../share/test_dataset/read_1.fasta ../share/test_dataset/read_2.fasta \
   --scafinfo ../share/test_dataset/scaf.info \
   --ref ../share/test_dataset/ref.fasta \
   --label pair scaf ref \
   --color "#0000ff" "#00ff00" "#ff0000" \
   -o output
</code>
</pre>



<a name="sec3.2"></a>
<h3>3.1. RNA scaffolder</h3>

<a name="sec3.2.1"></a>
<h4>3.1.1. Input </h4>

<p>
    SGTK takes as input contigs in FASTA format and paired-end RNA reads or single (unpaired) RNA-seq reads in FASTA and FASTQ.
<p>

    <a name="sec3.2.2"></a>
<h4>3.1.2. Command line options </h4>

<p>
    To run RNA scaffolder from the command line, type

<pre class="code">
<code>
   rna_scaffolder.py [options]
</code>
</pre>
Note that we assume that SGTK installation directory is added to the <code>PATH</code> variable (otherwise provide full path to SGTK executable: <code>&lt;installation dir>/rna_scaffolder.py</code>).

<a name="basicopt"></a>
<h4>Options</h4>

<p>
    <code>-c</code> (or <code>--contig</code>) <code> &lt;file_name> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;File with contigs in fasta format. Required option.
</p>

<p>
    <code>--rna-p &lt;file_name_1>  &lt;file_name_2> </code>
    &nbsp;&nbsp;&nbsp;&nbsp; File with left reads and file with right reads for paired-end library.
</p>

<p>
    <code>--rna-s &lt;file_name> </code>
    &nbsp;&nbsp;&nbsp;&nbsp; File for single-read library.
</p>

<p>
    <code>--gene_annotation &lt;file_name> </code>
    File with gene annotation for given contigs in gff format. Required option.
</p>

<p>
    <code>-o</code> (or <code>--local_output_dir</code>) <code> &lt;output_dir> </code><br>
    &nbsp;&nbsp;&nbsp;&nbsp;Specify the output directory. The default output directory is  "./"
</p>


<a name="sec3.2.3"></a>
<h4>3.2.3. Output</h4>

SGTK stores all output files in <code>&lt;output_dir> </code>, which is set by the user.

<ul>
    <li><code>&lt;output_dir>/scaffolds.fa</code> contains resulting scaffolds </li>
    <li><code>&lt;output_dir>/graph.gr</code> scaffold graph </li>
    <li><code>&lt;output_dir>/out.info</code> scaffolds description</li>
</ul>

<p>
    Scaffolds names in SGTK output FASTA files have the following format: <br><code>path179 </code><br> Here <code>179</code> is the number of the scaffold.
</p>


<a name="sec3.2.4"></a>
<h4>3.2.4. Example</h4>

<p>
    To test the toy data set, you can also run the following command from the SGTK <code>bin</code> directory:

<pre class="code">
<code>
   rna_scaffolder.py -c ../share/test_dataset/contigs.fasta --rna-p ../share/test_dataset/read_1.fasta \
   ../share/test_dataset/read_2.fasta --gene_annotation annotation.gff -o test
</code>
</pre>

<p>
    If you have several paired-end reads and unpaired reads, for example:

<p>
    <li> contigs
        <pre class="code">
   <code>
   contigs.fasta
   </code>
   </pre>

    <li> paired-end library 1
        <pre class="code">
<code>
   lib_pe1_left.fastq
   lib_pe1_right.fastq
</code>
</pre>
<p>
    <li> paired-end library 2

        <pre class="code">
<code>
   lib_pe2_left.fastq
   lib_pe2_right.fastq
</code>
</pre>

<p>
    <li> unpaired library 1
        <pre class="code">
<code>
   lib1.fastq
</code>
       </pre>

<p>
    <li> gene annotation
        <pre class="code">
<code>
   annotation.gff
</code>
</pre>
<p>

<pre class="code">
<code>
   rna_scaffolder.py -c contigs.fasta --rna-p lib_pe1_left.fastq lib_pe1_right.fastq \
   --rna-p lib_pe2_left.fastq --mp1-2 lib_pe2_right.fastq \
   --rna-s lib1.fastq --gene_annotation annotation.gff -o output
</code>
</pre>


<a name="sec4"></a>
<h2>4. References </h2>

<a name="sec5"></a>
<h2>5. Feedback and bug reports </h2>

<br/>
</body>
</html>

